File: test_p/bsp/mode_test.py
from bsp.log import my_log
from bsp.power_manage import power_ctrl
import os
import time
import subprocess
import threading
class mode_test_t:
    def __init__(self):
        self.thread_flag = True
        print(" ")
        self.exit_event = threading.Event()

    def __del__(self):
        print("mode test end ")
        self.thread_flag = False

    def usb_reset(self):
        power_ctrl.usb_set(0xff)
        time.sleep(0.3)
        power_ctrl.usb_set(0x00)
        print("usb reset")

    def process_test(self):

        r = os.popen('adb shell "ls /oemapp/bin"').read()
        print(r)
        file_list = r.strip().split('\n')
        print("文件列表:", file_list)
        for file_name in file_list:
            # 构造 adb shell 命令，检查进程是否存在
            command = f'adb shell "ps -ef | grep {file_name} | grep -v grep"'

            # 执行命令并获取结果
            process_check = os.popen(command).read().strip()

            # 判断进程是否在运行
            if process_check:
                # print(f"进程 '{file_name}' 正在运行:")
                print(process_check)
            else:
                my_log.error(f"'{file_name}' 进程未启动")
    def adb_connect(self):
        cnt = 20
        while (cnt) and not self.exit_event.is_set():
            time.sleep(2)
            cnt -=1
            if os.system("adb root") == 0:
                return 0
            self.usb_reset()
        return -1

    def reflash_test(self):
        my_log.debug("-> reflash test")
        if os.system("adb shell \"ls /proc/recoveryinfo\"") != 0:
            my_log.error("tsu_reflash err")
            return -1
        my_log.debug("Reflash test successful")
        return 0
    def cellular_test(self):

        my_log.debug("-> cellular test")
        if os.system("adb shell \"ps -ef| grep oemapp/bin/cellular | grep -v grep\"") != 0:
            my_log.error("cellular No startup")
            return -1

        # tsu_part_num_file = os.popen("adb shell \"cat /usrdata/conf/tsu_part_num.txt\"").read()
        # if(fgets(buf,sizeof(buf),fp)>(char*)0)
        
        #     my_log.debug("tsu_part_num: %s",buf)
        #     if(strncmp("8B100-3BWT-COMM-M1",buf,sizeof("8B100-3BWT-COMM-M1")-1))
            
        #         my_log.error("cat tsu_part_num.txt err != 8B100-3BWT-COMM-M1")
        #         return -1
        if os.system("adb shell \"ifconfig | grep rmnet_data\"") != 0:
            my_log.error("The dial status is wrong")
            return -1
        cert_cn_file = os.popen("adb shell \"cat /data/run_info/cert_cn.txt\"").read()
        my_log.debug(f"cert_cn : {cert_cn_file}")
        my_log.debug("cellular test successful")
        return 0
    def extract_rt_cnt(self):
        process_rt_dict = {}
        cmd = ['adb', 'shell', 'cat /data/run_info/process_monintor.txt']
        with subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True) as proc:
            lines = proc.stdout.readlines()
            
            for line in lines[1:]:
                line = line.strip()
                if not line:
                    continue
                columns = line.split()
                if len(columns) >= 2:
                    process_name = columns[0]
                    rt_cnt = columns[1] 
                    process_rt_dict[process_name] = rt_cnt  
        
        return process_rt_dict

    def reboot_test(self):
        
        my_log.debug("start reboot start")
        
        tm = 0
        if self.adb_connect() == -1:
            my_log.error("Adb connection failed")
            return -1
        adb_root_counter = 0
        
        while not self.exit_event.is_set():
            if os.system("adb root") != 0:
                self.usb_reset()
                adb_root_counter += 1
            if adb_root_counter >= 4:
                
                my_log.error("TSU has been shut down")
                adb_root_counter = 0
                # return -1
            time.sleep(2)
            tm += 1
            if tm % 149 == 0:
                process_rt_data  = self.extract_rt_cnt()
                for process, rt_cnt in process_rt_data.items():
                    if int(rt_cnt) > 1:
                        my_log.error(f"{process}: {rt_cnt}")



    def reboot_test_task_start(self):
        power_ctrl.cleanup()
        task = threading.Thread(target=self.reboot_test)
        task.start()

    def all_test(self):
        self.process_test()
        self.reflash_test()
        self.cellular_test()
    

mode_test = mode_test_t()

File: src/mode_test.cpp

#include "mode_test.h"
#include "adb.h"
#include "log.h"
#include "string.h"
#include "gpio.h"
#include <vector>
#include <cstdio>
/** */
#define POPEN_READ_LEN   128

static gl_status_t popen_data_get(std::string cmd)
{
    char buffer[512];
    FILE* fp = nullptr;
    fp = popen(cmd.c_str(),"r");
    while (fgets(buffer, sizeof(buffer), fp) != nullptr) {
         log_debug("%s",buffer);
    }
    pclose(fp);
    return STATE_OK;
}
/*time sync*/
gl_status_t timing_test()
{
    char buffer[256];
    FILE* fp = nullptr;
    info_printf("\n-> Time synchronization test\n");
    std::string cmd("adb shell \"cat /tmp/ql_time_sync_record_flag\"");
    if (system(cmd.c_str()))
    {
        log_error("/tmp/ql_time_sync_record_flag not exit");
        return STATE_FAIL;
    }

    popen_data_get(cmd);
    return STATE_OK;
}
/*REFLASH*/
gl_status_t reflash_test()
{
    info_printf("\n-> reflash test\n");
    if(system("adb shell \"ls /proc/recoveryinfo\"") != 0)
    {
        log_error("tsu_reflash err");
        return STATE_FAIL;
    }
    log_debug("Reflash test successful");
    return STATE_OK;
}

/*DAQ*/
gl_status_t daq_test()
{
    info_printf("\n-> daq test\n");
    if(system("adb shell \"ps | grep oemapp/bin/daq | grep -v grep\""))
    {
        log_error("DAQ No startup");
        return STATE_FAIL;
    }
    if(system("adb shell \"ls -l /usrdata/sqlite/daq_applied.db\""))
    {
        log_error("The daq applied.db file is missing");
    }
    if(system("adb shell \"rm -rf /usrdata/sqlite/upload/* \"") == 0)
    {
        
        gpio_write(GPIO_IG,1);
        sleep(2);
        gpio_write(GPIO_IG,0);
        sleep(30);
        if(system("adb shell \"ls -l /usrdata/sqlite/upload/ |grep .EOF\""))
        {
            log_error("DAQ is not working properly");
            return STATE_FAIL;
        }
    }

    log_debug("DAQ test successful");
    return STATE_OK;
}
/*ecall*/
gl_status_t ecall_test()
{
    char out[128] = {0};
    info_printf("\n-> ecall test\n");
    FILE *f = popen("adb shell \"cat /etc/ecall_version_*\"","r");
    if(fgets(out,128,f) > (char*)0)
    {
        log_debug("%s",out);
    }
    else
    {
        pclose(f);
        log_error("Unable to obtain ecall version");
        return STATE_FAIL;
    }
    pclose(f);
    return STATE_OK;
}
/*cellular*/
gl_status_t cellular_test()
{
    FILE *fp = NULL;
    char buf[128] = {0};
    info_printf("\n-> cellular test\n");
    if(system("adb shell \"ps | grep oemapp/bin/cellular | grep -v grep\""))
    {
        log_error("cellular No startup");
        return STATE_FAIL;
    }

    fp = popen("adb shell \"cat /usrdata/conf/tsu_part_num.txt\"","r");
    if(fgets(buf,sizeof(buf),fp)>(char*)0)
    {
        log_debug("tsu_part_num: %s",buf);
        if(strncmp("8B100-3BWT-COMM-M1",buf,sizeof("8B100-3BWT-COMM-M1")-1))
        {
            log_error("cat tsu_part_num.txt err != 8B100-3BWT-COMM-M1");
            return STATE_FAIL;
        }
    }

    if(system("adb shell \"ifconfig | grep rmnet_data\""))
    {
        log_error("The dial status is wrong");
        return STATE_FAIL;
    }

    memset(buf,0,sizeof(buf));
    FILE *f1 = popen("adb shell \"cat /data/run_info/cert_cn.txt\"","r");
    if(fgets(buf,sizeof(buf),f1))
    {
        log_debug("cert_cn: %s",buf);
    }
    pclose(f1);
    memset(buf,0,sizeof(buf));
    FILE *f2 = popen("adb shell \"cat /data/run_info/iccid.txt\"","r");
    if(fgets(buf,sizeof(buf),f2))
    {
        log_debug("iccid: %s",buf);
    }
    pclose(f2);
    memset(buf,0,sizeof(buf));
    FILE *f3 = popen("adb shell \"cat /data/run_info/imei.txt\"","r");
    if(fgets(buf,sizeof(buf),f3))
    {
        log_debug("imei: %s",buf);
    }
    pclose(f3);

    pclose(fp);
    log_debug("cellular test successful");
    return STATE_OK;
}